### PR TITLE
Merge request for issue #22 - apply BCs to operators

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

File: doc/dynamo0p3.rst [hunk 1]
```diff
@@ -1246,35 +1246,35 @@ where:
 Boundary Conditions
 -------------------
 
-In the dynamo0.3 API, boundary conditions for a field can be enforced
-by the algorithm developer by calling a particular Kernel called
-``enforce_bc_type``. This kernel takes a field as input and applies
+In the dynamo0.3 API, boundary conditions for a field or LMA operator can
```
Comment (arporter) [inline doc/dynamo0p3.rst]: At some point we should decide that this is no longer the 0.3 version of the API...maybe 1.0?

Comment (rupertford) [inline doc/dynamo0p3.rst]: Agreed, the problem is that it potentially involves a lot of changes as the dynamo0.3 name is used all over the place. I think we should rename it the LFRic1.0 API. We could possibly do this before the 1.4.0 release, but it would cause a delay.

File: src/tests/dynamo0p3_test.py [hunk 2]
```diff
@@ -2337,6 +2330,96 @@ def test_bc_kernel_field_only(monkeypatch):
...
+        # field rather than an operator
+        monkeypatch.setattr(arg, "_type", value="gh_field")
+        with pytest.raises(GenerationError) as excinfo:
+            _ = psy.gen
+        assert ("Expected a LMA operator from which to look-up boundary dofs "
+                "but kernel enforce_operator_bc_code has no such argument") \
+            in str(excinfo)
+
+
+def test_operator_bc_kernel_multi_args_err():  # pylint: disable=invalid-name
```
Comment (rupertford) [inline src/tests/dynamo0p3_test.py]: Shouldn't we also be testing that there is only one argument? This test implies it is valid for there to be more than one argument as long as only one of them is an operator.

File: src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90 [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  type(arg_type) :: meta_args(1) = (/                               &
+       arg_type(GH_OPERATOR, GH_INC, ANY_SPACE_1, ANY_SPACE_2)      &
+       /)
+  integer :: iterates_over = CELLS
+contains
+  procedure, nopass :: enforce_operator_bc_code
+end type
+
+!-------------------------------------------------------------------------------
+! Constenforce_operator_bcctors
```
Comment (rupertford) [inline src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90]: Are there typos here or is this really what is meant. If so, what does it mean?

**Comment by rupertford** [comment]
@arporter. First review completed.

File: src/tests/dynamo0p3_test.py (hunk 2 above)
Comment (arporter) [inline src/tests/dynamo0p3_test.py]: Good point. I've tightened up the code and associated tests. c772039

File: src/dynamo0p3.py [hunk 4]
```diff
@@ -2683,7 +2630,11 @@ def generate(self):
             # kernel (enforce_bc_kernel) arguments
             if self._kern.name.lower() == "enforce_bc_code" and \
                unique_fs.orig_name.lower() == "any_space_1":
-                self.bc_kernel(unique_fs)
+                self.field_bcs_kernel(unique_fs)
+        # Add boundary dofs array to the operator boundary condition
+        # kernel (enforce_operator_bc_kernel) arguments
+        if self._kern.name.lower() == "enforce_operator_bc_code":
+            self.operator_bcs_kernel(fs_to)
```
Comment (arporter) [inline src/dynamo0p3.py]: Good plan. I've done this in fce5d51. There is one sanity check just to make sure we have an operator to get the FS from. All others are within the `operator_bcs_kernel` routine. I've extended the tests to make sure we cover all of these checks.

File: src/dynamo0p3.py [hunk 5]
```diff
@@ -2683,7 +2630,11 @@ def generate(self):
             # kernel (enforce_bc_kernel) arguments
             if self._kern.name.lower() == "enforce_bc_code" and \
                unique_fs.orig_name.lower() == "any_space_1":
```
Comment (arporter) [inline src/dynamo0p3.py]: I don't know either. Maybe it was to help us trap any changes to the interface to `enforce_bc_code` that the Met Office didn't tell us about?

File: src/dynamo0p3.py [hunk 6]
```diff
@@ -2906,6 +2866,49 @@ def bc_kernel(self, function_space):
...
+        applies boundary conditions to a LMA operator '''
+        from f2pygen import DeclGen, AssignGen
+        self._arglist.append("boundary_dofs")
+        parent = self._parent
+        parent.add(DeclGen(parent, datatype="integer",
+                           pointer=True, entity_decls=[
+                               "boundary_dofs(:,:) => null()"]))
+        # Find the single, LMA operator argument that this kernel updates
+        # TODO replace this with args_filter() once the modification to
+        # make it a module function is on master
```
Comment (arporter) [inline src/dynamo0p3.py]: This is now redundant since you pointed out that we should only have 1 LMA operator argument anyway. I've put in sanity checks to ensure that's what we have.

File: src/dynamo0p3.py [hunk 7]
```diff
@@ -2906,6 +2866,49 @@ def bc_kernel(self, function_space):
...
+        op_args = []
```
Comment (arporter) [inline src/dynamo0p3.py]: Done (with checking).

File: src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90 (hunk 3 above)
Comment (arporter) [inline src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90]: Weird. It was a typo. Fixed now.

**Comment by arporter** [comment]
@rupertford I've updated the copyright/license notices. Ready for 2nd review...

File: src/dynamo0p3.py [hunk 8]
```diff
@@ -3240,42 +3262,36 @@ def operator_bcs_kernel(self, function_space):
...
-        for arg in self._kern.arguments.args:
-            if arg.type == "gh_operator":
-                op_args.append(arg)
-        # op_args = self.parent.args_filter(self._kern.arguments,
-        #                                   arg_types=VALID_OPERATOR_NAMES,
-        #                                   arg_accesses=["gh_write"])
-        if not op_args:
+        # Sanity checks - this kernel should only have a single LMA
+        # operator as argument
+        if len(self._kern.arguments.args) > 1:
```
Comment (rupertford) [inline src/dynamo0p3.py]: Should these sanity tests be in the base class. Are they note generally useful?

File: src/dynamo0p3.py [hunk 9]
```diff
@@ -2949,7 +2960,18 @@ def generate(self):
         # Add boundary dofs array to the operator boundary condition
         # kernel (enforce_operator_bc_kernel) arguments
         if self._kern.name.lower() == "enforce_operator_bc_code":
-            self.operator_bcs_kernel(fs_to)
+            # Minimal sanity check - this kernel should only have a single LMA
+            # operator as argument so the first one must be a gh_operator.
+            # Full sanity checking is done in the operator_bcs_kernel()
```
Comment (rupertford) [inline src/dynamo0p3.py]: I've commented on this below, but there are multiple implementations of the operator_bcs_kernel() in the different subclasses so there is no single implementation. I think the sanity checking should all be done here (in the base class).

File: src/tests/dynamo0p3_test.py [hunk 10]
```diff
@@ -2260,9 +2262,11 @@ def test_multi_kernel_specific():
...
     assert output3 not in generated_code
     # We only call enforce_bc if the field is on a vector space
-    output4 = '''IF (fs == w1 .or. fs == w2 .or. fs == w2h .or. fs == w2v) THEN
-        boundary_dofs => f1_proxy%vspace%get_boundary_dofs()
-      END IF'''
+    output4 = (
+        "IF (fs == w1 .or. fs == w2 .or. fs == w2h .or. fs == w2v .or. "
+        "fs == any_w2) THEN\n"
+        "        boundary_dofs => f1_proxy%vspace%get_boundary_dofs()\n"
+        "      END IF")
```
Comment (rupertford) [inline src/tests/dynamo0p3_test.py]: Very diligent to update tests for things that are not there :-)

File: src/tests/test_files/dynamo0p3/12.4_enforce_op_bc_kernel.f90 [hunk 11]
```diff
@@ -1,9 +1,42 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
+! However, it has been created with the help of the GungHo Consortium,
```
Comment (rupertford) [inline src/tests/test_files/dynamo0p3/12.4_enforce_op_bc_kernel.f90]: We don't need the GungHo bit as this was created after GungHo finished. If it wasn't we need the original GungHo license here instead (until we get agreement that we can take ownership).

File: src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90 [hunk 12]
```diff
@@ -1,3 +1,37 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
+! However, it has been created with the help of the GungHo Consortium,
```
Comment (rupertford) [inline src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90]: I think the original license should be at the top. If you have modified the code then adding BSD is good but could you please remove lines 5,6,7 and also add modified by A. Porter STFC.

**Comment by rupertford** [comment]
@arporter back to you Andy

File: examples/dynamo/eg4/solver_mod.x90 [hunk 13]
```diff
@@ -33,6 +33,11 @@
 ! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
+! (c) The copyright relating to this work is owned jointly by the Crown,
+! Met Office and NERC 2014.
+! However, it has been created with the help of the GungHo Consortium,
```
Comment (arporter) [inline examples/dynamo/eg4/solver_mod.x90]: Agreed. I have tweaked them slightly so have added the necessary text and re-ordered the license text.

File: src/dynamo0p3.py [hunk 14]
```diff
@@ -2901,7 +2911,8 @@ def generate(self):
                     self.stencil(arg)
             elif arg.type == "gh_operator":
                 self.operator(arg)
-                fs_to = arg.function_space_to
+            elif arg.type == "gh_columnwise_operator":
+                self.cma_operator(arg)
```
Comment (arporter) [inline src/dynamo0p3.py]: Good point. Done.

File: src/dynamo0p3.py (hunk 9 above)
Comment (arporter) [inline src/dynamo0p3.py]: I've just remembered why I did it this way - I was concerned that the implementation of each method should stand on its own two feet and validate its arguments. If we pull the validation checks out and put them in the base class that calls them then, should we ever add additional calls to them from somewhere else, we will lose the validation checks. On the other hand I suspect we are unlikely to call them from elsewhere and we can add comments documenting the checks. A third possibility is to put some sort of `validate` method in the base class and call that to do the checks.

Comment (arporter) [inline src/dynamo0p3.py]: Having talked this through with Rupert we have agreed that methods such as `operator_bcs_kernel()` should be private. I've created ticket #36 for this.
I'm now happy with putting the checks into `generate()` so will do that.

File: src/tests/test_files/dynamo0p3/12.4_enforce_op_bc_kernel.f90 (hunk 11 above)
Comment (arporter) [inline src/tests/test_files/dynamo0p3/12.4_enforce_op_bc_kernel.f90]: OK. It is post GHo so removed.

File: src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90 (hunk 12 above)
Comment (arporter) [inline src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90]: Done.

**Comment by arporter** [comment]
Sanity checks moved as requested. All still covered by test suite. pep8 and pylint are OK. I've updated the license text as requested.
Back to @rupertford ...

File: examples/dynamo/eg4/enforce_bc_kernel_mod.f90 [hunk 15]
```diff
@@ -1,3 +1,37 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017, Science and Technology Facilities Council
+! However, it has been created with the help of the GungHo Consortium,
```
Comment (rupertford) [inline examples/dynamo/eg4/enforce_bc_kernel_mod.f90]: Great.

File: examples/dynamo/eg4/solver_mod.x90 (hunk 13 above)
Comment (rupertford) [inline examples/dynamo/eg4/solver_mod.x90]: Great.

File: src/dynamo0p3.py (hunk 14 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great.

File: src/dynamo0p3.py (hunk 9 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great, this is the right solution IMHO.

File: src/dynamo0p3.py (hunk 8 above)
Comment (rupertford) [inline src/dynamo0p3.py]: This has been done (see earlier comment). Great.

File: src/tests/test_files/dynamo0p3/12.4_enforce_op_bc_kernel.f90 (hunk 11 above)
Comment (rupertford) [inline src/tests/test_files/dynamo0p3/12.4_enforce_op_bc_kernel.f90]: Great. I've added you as author as well.

File: src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90 (hunk 12 above)
Comment (rupertford) [inline src/tests/test_files/dynamo0p3/enforce_operator_bc_kernel_mod.F90]: Great.

**Comment by rupertford** [comment]
latest changes are pep8, pyflakes and pylint clean
examples run
documentation builds
branch is up to date
tests pass or xfail
coverage checking done previously
Good job.
Pull request approved.
